src/subject_name/dns_name.rs

```diff
@@ -130,24 +122,145 @@ impl<'a> DnsNameRef<'a> {
     }
 }
 
-/// Requires the `alloc` feature.
-#[cfg(feature = "alloc")]
 impl core::fmt::Debug for DnsNameRef<'_> {
     fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
-        let lowercase = self.clone().to_owned();
-        f.debug_tuple("DnsNameRef").field(&lowercase.0).finish()
+        f.write_str("DnsNameRef(\"")?;
+
+        // Convert each byte of the underlying ASCII string to a `char` and
+        // downcase it prior to formatting it. We avoid self.clone().to_owned()
+        // since it requires allocation.
+        for &ch in self.0 {
+            core::fmt::Display::fmt(&char::from(ch).to_ascii_lowercase(), f)?;
+        }
+
+        f.write_str("\")")
     }
 }
 
-#[cfg(not(feature = "alloc"))]
-impl core::fmt::Debug for DnsNameRef<'_> {
+impl<'a> From<DnsNameRef<'a>> for &'a str {
+    fn from(DnsNameRef(d): DnsNameRef<'a>) -> Self {
+        // The unwrap won't fail because DnsNameRefs are guaranteed to be ASCII
+        // and ASCII is a subset of UTF-8.
+        core::str::from_utf8(d).unwrap()
+    }
+}
+
+/// A DNS Name suitable for use in the TLS Server Name Indication (SNI)
+/// extension and/or for use as the reference hostname for which to verify a
+/// certificate.
+pub enum GeneralDnsNameRef<'name> {
```

Comment:
I'm curious, I don't see a direct justification for why we need/want this intermediate layer of `GeneralDnsName`, splitting out `WildcardDnsName`s from the other kind. Is that something that happens in other parts of the code implicitly and here we need/want to make it explicit?

(The docstring here doesn't really talk about this aspect at all.)

Reply:
I'll see if I can reverse engineer the reasoning here and get back to you. I don't have an answer on hand.

Reply:
I did some spelunking through the history of this work (it goes all the way back to 2017!) to figure out an answer. I'm not sure I have the full picture in my head but made some progress.

The original motivation appears to be to have the type system capable of distinguishing between whether a "presented DNS ID" is a DNS name ref, or a wildcard. The general idea for this was articulated in https://github.com/briansmith/webpki/issues/66 as a way to tighten up `GeneralName::DnsName(untrusted::Input<'a>)` (_which isn't done in this PR, but perhaps we should consider doing this in a follow-up?_). In particular it seems like the notion of a "DNS name reference" precluding wildcards follows from [RFC 6125](https://www.rfc-editor.org/rfc/rfc6125) (there's explicit discussion of wildcards in [§7.2](https://www.rfc-editor.org/rfc/rfc6125#section-7.2)).

When the topic of adding the functionality to collect DNS names from a cert was raised Brian [advocated](https://github.com/briansmith/webpki/pull/65#issuecomment-425305373) in [PR 65](https://github.com/briansmith/webpki/pull/65) that it should be done with that type system improvement in mind. Later in [PR 91](https://github.com/briansmith/webpki/pull/91) he also said that his support for including wildcards was [conditional on this type system distinction](https://github.com/briansmith/webpki/pull/91#issuecomment-475790033).

WYDT? Is there a change I could make in this branch to help clarify this, maybe with a reference in the doc strings? It seems like a worthwhile distinction to maintain.

Reply:
Enforcing that presented DNS IDs may contain a wildcard while DNS name references may not is also [handled similarly in NSS libpkix](https://github.com/nss-dev/nss/blob/df545b17d26e945a89437af035119a564ad79cdd/lib/mozpkix/lib/pkixnames.cpp#L1883-L2065).

Reply:
I made an attempt to capture the reason for the type system distinction in the comment on `GeneralDnsNameRef`. I also updated the `WildcardDnsNameRef` comment to remove mention of SNI (which shouldn't contain a wildcard), and to mention that we're stricter than RFC 6125 when it comes to processing wildcard labels (only allowing a bare wildcard label). Lastly I updated `DnsNameRef` and `DnsName` to explicitly mention the absence of wildcards. Hopefully that helps!